lib/std/crypto/keccak_p.zig

```diff
@@ -0,0 +1,241 @@
+                self.st[i / @sizeOf(T)] = mem.readIntLittle(T, padded[0..]);
+            }
+        }
+
+        /// XOR a byte into the state at a given offset.
+        pub fn addByte(self: *Self, byte: u8, offset: usize) void {
+            const z = @sizeOf(T) * @truncate(math.Log2Int(T), offset % @sizeOf(T));
+            self.st[offset / @sizeOf(T)] ^= @as(T, byte) << z;
+        }
+
+        /// XOR bytes into the beginning of the state.
+        pub fn addBytes(self: *Self, bytes: []const u8) void {
+            var i: usize = 0;
+            while (i + @sizeOf(T) <= bytes.len) : (i += @sizeOf(T)) {
+                self.st[i / @sizeOf(T)] ^= mem.readIntLittle(T, bytes[i..][0..@sizeOf(T)]);
+            }
+            if (i < bytes.len) {
+                var padded = [_]u8{0} ** @sizeOf(T);
+                mem.copy(u8, padded[0 .. bytes.len - i], bytes[i..]);
+                self.st[i / @sizeOf(T)] ^= mem.readIntLittle(T, padded[0..]);
+            }
+        }
+
+        /// Extract the first bytes of the state.
+        pub fn extractBytes(self: *Self, out: []u8) void {
+            var i: usize = 0;
+            while (i + @sizeOf(T) <= out.len) : (i += @sizeOf(T)) {
+                mem.writeIntLittle(T, out[i..][0..@sizeOf(T)], self.st[i / @sizeOf(T)]);
+            }
+            if (i < out.len) {
+                var padded = [_]u8{0} ** @sizeOf(T);
+                mem.writeIntLittle(T, padded[0..], self.st[i / @sizeOf(T)]);
+                mem.copy(u8, out[i..], padded[0 .. out.len - i]);
+            }
+        }
+
+        /// XOR the first bytes of the state into a slice of bytes.
+        pub fn xorBytes(self: *Self, out: []u8, in: []const u8) void {
+            assert(out.len == in.len);
+
+            var i: usize = 0;
+            while (i + @sizeOf(T) <= in.len) : (i += @sizeOf(T)) {
+                const x = mem.readIntNative(T, in[i..][0..@sizeOf(T)]) ^ mem.nativeToLittle(T, self.st[i / @sizeOf(T)]);
+                mem.writeIntNative(T, out[i..][0..@sizeOf(T)], x);
+            }
+            if (i < in.len) {
+                var padded = [_]u8{0} ** @sizeOf(T);
+                mem.copy(u8, padded[0 .. in.len - i], in[i..]);
+                const x = mem.readIntNative(T, &padded) ^ mem.nativeToLittle(T, self.st[i / @sizeOf(T)]);
+                mem.writeIntNative(T, &padded, x);
+                mem.copy(u8, out[i..], padded[0 .. in.len - i]);
+            }
+        }
+
+        /// Set the words storing the bytes of a given range to zero.
+        pub fn clear(self: *Self, from: usize, to: usize) void {
+            mem.set(T, self.st[from / @sizeOf(T) .. (to + @sizeOf(T) - 1) / @sizeOf(T)], 0);
+        }
+
+        /// Clear the entire state, disabling compiler optimizations.
+        pub fn secureZero(self: *Self) void {
+            std.crypto.utils.secureZero(T, &self.st);
+        }
+
+        inline fn round(self: *Self, rc: T) void {
```

Comment:
While you're making big changes to the package, it might be worthwhile to consider doing f1600 differently. By reordering the operations a bit, and grouping rounds, it can be quite a bit faster.

See, eg. https://github.com/golang/crypto/blob/ebe92624d1428c68f92576e1d27cc65d62bc2f7e/sha3/keccakf.go#L51 which groups 4 rounds or https://github.com/pq-crystals/kyber/blob/518de2414a85052bb91349bcbcc347f391292d5b/ref/fips202.c#L126 which groups 2.

Reply:
This would be closer to how you would write it in Zig: https://github.com/cloudflare/circl/blob/main/simd/keccakf1600/internal/asm/src.go#L13

Reply:
The fastest implementation from XKCP even groups 6 rounds. We can certainly experiment later to see what the gains would be. 

But let's get SHAKE merged first :)

Reply:
> The fastest implementation from XKCP even groups 6 rounds. 

In microbenchmarks. In larger benchmarks it might do worse because of cache pressure.